pkg/platform/api/buildplanner/model/buildexpression.go

```diff
@@ -97,122 +167,254 @@ type BuildExpression struct {
 //	  }
 //	}
 func NewBuildExpression(data []byte) (*BuildExpression, error) {
-	expression := make(map[string]interface{})
-	err := json.Unmarshal(data, &expression)
+	m := make(map[string]interface{})
+	err := json.Unmarshal(data, &m)
 	if err != nil {
-		return nil, errs.Wrap(err, "Could not unmarshal JSON")
+		return nil, errs.Wrap(err, "Could not unmarshal build expression")
 	}
 
-	solveNode, err := getSolveNode(expression)
-	if err != nil {
-		return nil, errs.Wrap(err, "Could not get solve node")
+	letValue, ok := m["let"]
+	if !ok {
+		return nil, errs.New("Build expression has no 'let' key")
+	}
+	letMap, ok := letValue.(map[string]interface{})
+	if !ok {
+		return nil, errs.New("'let' key is not a JSON object")
 	}
 
-	requirements, err := getRequirements(solveNode)
+	ctx := context{}
+	let, err := newLet(ctx, letMap)
 	if err != nil {
-		return nil, errs.Wrap(err, "Could not get requirements node")
+		return nil, errs.Wrap(err, "Could not parse 'let' key")
 	}
 
-	return &BuildExpression{
-		expression:   expression,
-		solveNode:    &solveNode,
-		requirements: requirements,
-	}, nil
-}
-
-// validateRequirements ensures that the requirements in the BuildExpression contain
-// both the name and namespace fields. These fileds are used for requirement operations.
-func validateRequirements(requirements []interface{}) error {
-	for _, requirement := range requirements {
-		r, ok := requirement.(map[string]interface{})
-		if !ok {
-			return errs.New("Requirement in BuildExpression is malformed")
-		}
+	expr := &BuildExpression{let}
 
-		_, ok = r[RequirementNameKey]
-		if !ok {
-			return errs.New("Requirement in BuildExpression is missing name field: %#v", r)
-		}
-		_, ok = r[RequirementNamespaceKey]
-		if !ok {
-			return errs.New("Requirement in BuildExpression is missing namespace field: %#v", r)
-		}
+	err = expr.validateRequirements()
+	if err != nil {
+		return nil, errs.Wrap(err, "Could not validate requirements")
 	}
-	return nil
-}
 
-// Requirements returns the requirements in the BuildExpression.
-func (bx BuildExpression) Requirements() []Requirement {
-	return bx.requirements
+	return expr, nil
 }
 
-// Update updates the BuildExpression's requirements based on the operation and requirement.
-func (bx *BuildExpression) Update(operation Operation, requirement Requirement, timestamp strfmt.DateTime) error {
-	var err error
-	switch operation {
-	case OperationAdded:
-		err = bx.addRequirement(requirement)
-	case OperationRemoved:
-		err = bx.removeRequirement(requirement)
-	case OperationUpdated:
-		err = bx.updateRequirement(requirement)
-	default:
-		return errs.New("Unsupported operation")
+func newLet(ctx context, m map[string]interface{}) (*Let, error) {
+	ctx.push(ctxLet)
+	defer ctx.pop()
```

Comment:
Technically you're ignoring an error if the stack is empty. While I don't see how this could fail in this function, improper stack management in other functions may cause an error. Perhaps an error check is warranted.

Reply:
Yes, you're right however I don't think the `pop` function needs to return an error or a string. We really only want to inspect what's in the stack with `contains`